File: backend/app/src/main/kotlin/net/animeta/backend/controller/v2/PostController.kt
package net.animeta.backend.controller.v2

import net.animeta.backend.dto.PostDTO
import net.animeta.backend.exception.ApiException
import net.animeta.backend.repository.HistoryRepository
import net.animeta.backend.repository.RecordRepository
import net.animeta.backend.serializer.PostSerializer
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/v2/posts/{id}")
class PostController(val historyRepository: HistoryRepository,
                     val recordRepository: RecordRepository,
                     val postSerializer: PostSerializer) {
    @GetMapping
    fun get(@PathVariable id: Int,
            @RequestParam(required = false) options: PostSerializer.Options?): PostDTO {
        val history = historyRepository.findById(id).orElseThrow { ApiException.notFound() }
        return postSerializer.serialize(history, options ?: PostSerializer.legacyOptions(includeRecord = true, includeUser = true))
    }
}

File: backend-ts/src/controllers/post.controller.ts
import { Controller, DefaultValuePipe, Get, Param, ParseIntPipe, Query } from "@nestjs/common";
import { PostDTO } from 'shared/types';
import { InjectRepository } from "@nestjs/typeorm";
import { Repository } from "typeorm";
import { ApiException } from "./exceptions";
import { History } from "src/entities/history.entity";
import { PostSerializer, PostSerializerOptions } from "src/serializers/post.serializer";

@Controller('/api/v4/posts/:id')
export class PostController {
  constructor(
    @InjectRepository(History) private historyRepository: Repository<History>,
    private postSerializer: PostSerializer,
  ) {}

  @Get()
  async get(
    @Param('id', new ParseIntPipe()) id: number,
    @Query('options') optionsJson: string | null,
  ): Promise<PostDTO> {
    const history = await this.historyRepository.findOne(id)
    if (!history) throw ApiException.notFound()
    const options: PostSerializerOptions =
      optionsJson ? JSON.parse(optionsJson) : PostSerializer.legacyOptions({
        includeRecord: true,
        includeUser: true,
      })
    return this.postSerializer.serialize(history, options)
  }
}
